notation.go

```diff
@@ -63,31 +41,202 @@ type SignOptions struct {
+
+	// Action is the intended action for the given verification type as defined
+	// in the trust policy
+	Action trustpolicy.ValidationAction
+
+	// Error is set if there are any errors during the verification process
+	Error error
+}
+
+// VerificationOutcome encapsulates a signature blob's descriptor, its content,
+// the verification level and results for each verification type that was
+// performed.
+type VerificationOutcome struct {
+	// SignatureBlobDescriptor is descriptor of the signature envelope blob
+	SignatureBlobDescriptor *ocispec.Descriptor
+
+	// EnvelopeContent contains the details of the digital signature and
+	// associated metadata
+	EnvelopeContent *signature.EnvelopeContent
+
+	// VerificationLevel describes what verification level was used for
+	// performing signature verification
+	VerificationLevel *trustpolicy.VerificationLevel
+
+	// VerificationResults contains the verifications performed on the signature
+	// and their results
+	VerificationResults []*ValidationResult
+
+	// SignedAnnotations contains arbitrary metadata relating to the target
+	// artifact that was signed
+	SignedAnnotations map[string]string
+
+	// Error that caused the verification to fail (if it fails)
+	Error error
 }
 
 // Verifier is a generic interface for verifying an artifact.
 type Verifier interface {
-	// Verify verifies the signature and returns the verified descriptor and
-	// metadata of the signed artifact.
-	Verify(ctx context.Context, signature []byte, opts VerifyOptions) (Descriptor, error)
+	// Verify verifies the signature blob and returns the artifact
+	// descriptor upon successful verification.
+	Verify(ctx context.Context, signature []byte, opts VerifyOptions) (ocispec.Descriptor, *VerificationOutcome, error)
 }
 
-// Service combines the signing and verification services.
-type Service interface {
-	Signer
-	Verifier
+// Verify performs signature verification on each of the notation supported
+// verification types (like integrity, authenticity, etc.) and return the
+// verification outcomes.
+// For more details on signature verification, see
+// https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#signature-verification
+func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, opts VerifyOptions) (ocispec.Descriptor, []*VerificationOutcome, error) {
+	var verificationOutcomes []*VerificationOutcome
+	artifactRef := opts.ArtifactReference
+	// passing nil signature to check 'skip'
+	_, outcome, err := verifier.Verify(ctx, nil, opts)
+	if outcome == nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+	verificationLevel := outcome.VerificationLevel
+	if verificationLevel.Name == trustpolicy.LevelSkip.Name {
+		verificationOutcomes = append(verificationOutcomes, &VerificationOutcome{VerificationLevel: verificationLevel})
+		return ocispec.Descriptor{}, verificationOutcomes, nil
+	}
+
+	// get signature manifests
+	var success bool
```

Comment:
It is still needed, the final step of Verify() function checks if success is true. If it's true we will `return targetArtifactDesc, verificationOutcomes, nil`.

Reply:
Why do we need to return targetArtifactDesc? The `VerificationOutcome` contains `EnvelopeContent` which will have descriptor that is in payload.

https://github.com/patrickzheng200/notation-go/blob/af27f20064cbc623058cd201102dfbbc8b74837a/notation.go#L112

Reply:
The Payload.Content is a []byte instead of a descriptor.

Reply:
We will do `json.Unmarshal(outcome.EnvelopeContent.Payload.Content, payload)` only when there is successful verification and IMO its okay to do some repetitive work in favor of readability.

Reply:
In the most scenarios, only the descriptor of the signed artifact is interested. The `Verify()` function will be invoked as
```golang
desc, _, err := Verify(...)
if err != nil {
    return err
}
return downloadAndProcess(desc)
```
This design is to make the notation-go SDK more user friendly and easier to use.

Reply:
Instead of a descriptor should we return a signature envelope that would contain the descriptor and other data?
The usecase where a signature envelope would be useful is returning the user-defined signed metadata as part of verification.

Reply:
> Instead of a descriptor should we return a signature envelope that would contain the descriptor and other data? The usecase where a signature envelope would be useful is returning the user-defined signed metadata as part of verification.

Oh yeah, for the signature envelope blob, I added a field called `RawSignature` inside VerificationOutcome. It's also included in this PR.

Reply:
Note that `RawSignature` is better than a signature envelope type since a bad signature cannot even be parsed.